improvement(full-scan): Add support for reversed queries

yarongilor opened the pull request:
Added an option to existing full-scan stress of issuing a reversed query.
Trello: https://trello.com/c/xnqJRs4n

As for adding a verification for the reversed query output, it may be flaky since requires 2 different queries at 2 different point in time, thus it is not guaranty that data is not modified in between.
## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

yarongilor commented:
Tested in: https://jenkins.scylladb.com/job/scylla-staging/job/Longevity_yaron/job/longevity-large-partition-asymmetric-cluster-3h_reversed_queries/

roydahan commented on sdcm/cluster.py:
```diff
@@ -3611,6 +3611,26 @@ def verify_node_setup(start_time):
+    :return:
+    """
+    pks_list = []
+    cmd = f'select {pk_name} from {ks_cf} limit {num_of_pk}'
```
Why with cqlsh and not with driver?
Please change

yarongilor replied:
@roydahan , why not? it is just a small "internal" query

roydahan replied:
because everything that we can should be done through driver.

yarongilor replied:
fixed.

roydahan commented on sdcm/cluster.py:
```diff
@@ -3611,6 +3611,26 @@ def verify_node_setup(start_time):
     return wrapper
 
 
+def get_partition_keys(ks_cf: str, db_node: BaseNode, pk_name: str = 'pk', num_of_pk: int = 100) -> List[str]:
```
Why on cluster.py and not utils or somewhere more general.

yarongilor replied:
because it uses 'BaseNode' which is found in cluster.py..
(so now cluster.py has 5482 lines and definitelly needs a general refactor task)

fgelcer replied:
well, that is not a reason, because you can pass the `db_node` parameter to that function, anywhere else in the code

yarongilor replied:
This will cause a cyclic import..

soyacz replied:
so how about moving out 'BaseNode' out of cluster.py to some 'node.py'? Shouldn't be hard.

yarongilor replied:
That could be a good idea for a separate task.

yarongilor replied:
fixed eventually.

roydahan commented on test-cases/longevity/longevity-large-partition-3h.yaml:
```diff
@@ -40,3 +40,5 @@ nemesis_interval: 15
 
 space_node_threshold: 644245094
+
+run_fullscan: 'scylla_bench.test, 10, allow_reversed_queries' # 'ks.cf|random, interval(min)''
```
Why did you add it only to this longevity?
Please add to all large partitions longevities.

Now the problem is that it’s random and we can’t control it like this.
So if one want to test this feature he needs to wait and hope to get it running enough times.
Please change it so we can control it.

roydahan commented:
@yarongilor 
Please remove it from the full scan implementation and create a new parameter specifically for it with its own thread.
The interval should be more frequent (possibly every 1 min).
Add it to all large partition longevities, for the Scylla bench table (not random).

yarongilor commented:
Tests passed ok with the new reversed queries full-scans:
```
< t:2021-11-23 20:19:26,864 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-11-23 20:19:26.861: (FullScanEvent Severity.NORMAL) period_type=begin event_id=862a5e81-0124-4aec-8d94-d99ed644fbde node=longevity-large-partitions-3h-rever-db-node-071db80c-2 select_from=scylla_bench.test
< t:2021-11-23 20:19:28,750 f:full_scan_thread.py l:92   c:FullScanThread       p:INFO  > Will run command "select * from scylla_bench.test where pk = 111 order by ck desc USING TIMEOUT 120s"
< t:2021-11-23 20:19:29,003 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-11-23 20:19:28.998: (FullScanEvent Severity.NORMAL) period_type=end event_id=862a5e81-0124-4aec-8d94-d99ed644fbde duration=2s node=longevity-large-partitions-3h-rever-db-node-071db80c-2 select_from=scylla_bench.test message=full scan ended successfully
```
Following above comments, it now requires a wider refactor of full-scan thread.

yarongilor commented:
> @yarongilor Please remove it from the full scan implementation and create a new parameter specifically for it with its own thread. The interval should be more frequent (possibly every 1 min). Add it to all large partition longevities, for the Scylla bench table (not random).

done

yarongilor commented on sdcm/cluster.py:
```diff
@@ -3611,6 +3611,26 @@ def verify_node_setup(start_time):
     return wrapper
 
 
+def get_partition_keys(ks_cf: str, db_node: BaseNode, pk_name: str = 'pk', num_of_pk: int = 100) -> List[str]:
```
https://github.com/scylladb/scylla-cluster-tests/pull/4147#discussion_r756314243

yarongilor commented on sdcm/cluster.py:
```diff
@@ -3611,6 +3611,26 @@ def verify_node_setup(start_time):
+    Return list of partitions from a requested table
+    :param ks_cf:
+    :param db_node:
+    :param num_of_pk:
```
fixed.

yarongilor commented on sdcm/full_scan_thread.py:
```diff
@@ -16,23 +17,30 @@
         self.log = logging.getLogger(self.__class__.__name__)
         self._thread = threading.Thread(daemon=True, name=self.__class__.__name__, target=self.run)
 
+    def generate_query_options(self) -> dict:
```
removed.

yarongilor commented:
Tested ok in: https://jenkins.scylladb.com/job/scylla-staging/job/Longevity_yaron/job/longevity-large-partition-asymmetric-cluster-3h_reversed_queries/9/execution/node/142/log/ 
```
< t:2021-11-25 10:17:44,682 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-11-25 10:17:44.679: (ReversedQueryEvent Severity.NORMAL) period_type=begin event_id=795c5c08-4b70-4aea-8e12-630f40035680 node=longevity-large-partitions-3h-rever-db-node-6e00887e-2 select_from=scylla_bench.test
< t:2021-11-25 10:17:45,436 f:read_operation_thread.py l:81   c:ReversedQueryThread  p:INFO  > Will run command "select * from scylla_bench.test where pk = 111 order by ck desc USING TIMEOUT 300s"
< t:2021-11-25 10:17:45,584 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-11-25 10:17:45.582: (ReversedQueryEvent Severity.NORMAL) period_type=end event_id=795c5c08-4b70-4aea-8e12-630f40035680 duration=0s node=longevity-large-partitions-3h-rever-db-node-6e00887e-2 select_from=scylla_bench.test message=ReversedQueryEvent operation ended successfully

```

```
< t:2021-11-25 09:57:28,035 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-11-25 09:57:28.021: (FullScanEvent Severity.NORMAL) period_type=begin event_id=3c2892ad-e579-4e40-86b3-3038e8462f09 node=longevity-large-partitions-3h-rever-db-node-6e00887e-5 select_from=scylla_bench.test
< t:2021-11-25 09:59:04,138 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-11-25 09:59:04.126: (FullScanEvent Severity.WARNING) period_type=end event_id=3c2892ad-e579-4e40-86b3-3038e8462f09 duration=1m36s node=longevity-large-partitions-3h-rever-db-node-6e00887e-5 select_from=scylla_bench.test message=Error from server: code=1200 [Coordinator node timed out waiting for replica nodes' responses] message="Operation timed out for scylla_bench.test - received only 0 responses from 1 CL=ONE." info={'consistency': 'ONE', 'required_responses': 1, 'received_responses': 0}
< t:2021-11-25 10:09:04,239 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-11-25 10:09:04.236: (FullScanEvent Severity.NORMAL) period_type=begin event_id=a8f40495-e9d2-4a31-be3c-009511db1205 node=longevity-large-partitions-3h-rever-db-node-6e00887e-4 select_from=scylla_bench.test

```

fgelcer commented on longevity_test.py:
```diff
@@ -242,11 +241,13 @@ def test_custom_time(self):
-            self.log.info('Fullscan target: {} Fullscan interval: {}'.format(fullscan['ks.cf'],
-                                                                             fullscan['interval']))
-            self.run_fullscan_thread(ks_cf=fullscan['ks.cf'], interval=fullscan['interval'])
+        fs_params = self._get_read_operation_params(read_operation='run_fullscan')
```
i think it is better to use a "full" meaningful name.. what about `full_scan_params`?

yarongilor replied:
fixed

fgelcer commented on longevity_test.py:
```diff
@@ -242,11 +241,13 @@ def test_custom_time(self):
+        if fs_params:
+            self.run_fullscan_thread(ks_cf=fs_params['ks.cf'], interval=fs_params['interval'])
+
+        rq_params = self._get_read_operation_params(read_operation='run_reversed_query')
```
same here, what about `reverse_query_params`?

yarongilor replied:
fixed

fgelcer reviewed (commented):
just wanted to be sure it will be super clear to future readers of this code, but other than that, LGTM

roydahan commented on sdcm/sct_config.py:
```diff
@@ -793,6 +793,8 @@ class SCTConfiguration(dict):
         dict(name="run_fullscan", env="SCT_RUN_FULLSCAN", type=str,
              help=""),
+        dict(name="run_reversed_query", env="SCT_RUN_REVERSED_QUERY", type=str,
+             help=""),
```
Please add information in the "help"

yarongilor replied:
fixed

roydahan commented on sdcm/read_operation_thread.py:
```diff
@@ -50,21 +51,24 @@ def randomly_add_timeout(cmd) -> str:
 
-    def run_fullscan(self, db_node: BaseNode):  # pylint: disable=too-many-locals
-        ks_cf = self.get_ks_cs(db_node)
+    def run_read_operation(self, read_operation_event):  # pylint: disable=too-many-locals
```
The name "read_operation" that you changed everywhere, isn't clear.
Change it back to fullscan.
The reversed query is a partition full scan, so you can treat it as such.
We should have "full_partition_scan" and "full_partition_scan_reveresed_order" (aka revresed_query).

yarongilor replied:
@roydahan , how about renaming to just "scan_operation"? it may be less confusing

roydahan replied:
ok.

yarongilor replied:
fixed

roydahan commented on sdcm/read_operation_thread.py:
```diff
@@ -81,25 +85,64 @@ def run_fullscan(self, db_node: BaseNode):  # pylint: disable=too-many-locals
+
+
+class ReversedQueryThread(ReadOperationThread):
+    reversed_query_suffix = ' where pk = {} order by ck desc'
```
This is not enough.
As I wrote you in the trello task, it should support other options like:
- ck limitations (e.g. ck > {}, ck >{} and ck < {}, ...).
- Limit (e.g .... order by ck desc LIMIT 100)
- Paging (e.g. retrive only X pages or retrive all).

Every query that you build, should have its matching "opposite" query (notice that it's not always simple because for example when you limit 100, you should retrieve all but take only the last 100).
This "opposite" query should be used to verify the result we got from the reversed query.
Make this comparison optional for now (controlled through the yaml params of run_revresed_query") because I don't know how well it behaves and if the results will be completely accurate.
(You'll need to test it).
In any case, if the results doesn't match, you should raise an ERROR event.

yarongilor replied:
fixed.

roydahan commented on sdcm/read_operation_thread.py:
```diff
@@ -81,25 +85,64 @@ def run_fullscan(self, db_node: BaseNode):  # pylint: disable=too-many-locals
+
+
+class ReversedQueryThread(ReadOperationThread):
+    reversed_query_suffix = ' where pk = {} order by ck desc'
```
pk and ck can't be hardcoded, these are the names only for scylla_bench.test.
If later we want to test it on different table, it wouldn't work!

yarongilor replied:
fixed

roydahan reviewed (changes requested):
Please address EVERY COMMENT I left below and comment on it when addressing it.

fgelcer commented:
@yarongilor , please resolve conflicts

what is its status?
did you start a run with the packages we got from Piotr?

roydahan commented on sdcm/scan_operation_thread.py:
```diff
@@ -81,25 +86,76 @@ def run_fullscan(self, db_node: BaseNode):  # pylint: disable=too-many-locals
+    # Should support the following query options:  # TODO: finish implementing
+    # 1) ck = ?
+    # 2) ck > ?
+    # 3) ? < ck < ?
```
Are you sure this is the supported format?

yarongilor replied:
fixed

roydahan commented on test-cases/longevity/longevity-large-partition-200k_pks-4days.yaml:
```diff
@@ -68,3 +68,5 @@ user_prefix: 'longevity-large-partitions-200k-pks-4d'
 
 stop_test_on_stress_failure: false
+
+run_full_partition_scan: '{"ks_cf": "scylla_bench.test", "interval": 2, "pk_name":"pk", "rows_count": 100000}' # 'ks.cf, interval(min), partition-key name, number-of-rows-per-partition'
```
What  is the rows_count used for?

yarongilor replied:
its number-of-rows-per-partition and is used for taking a random number out of it in order to filter the ck.

fgelcer commented on test-cases/longevity/longevity-large-partition-4days.yaml:
```diff
@@ -24,3 +24,5 @@ space_node_threshold: 644245094
 table_name: "scylla_bench.test"
 primary_key_column: "pk"
+
+run_full_partition_scan: '{"ks_cf": "scylla_bench.test", "interval": 2, "pk_name":"pk", "rows_count": 10000000}' # 'ks.cf, interval(min), partition-key name, number-of-rows-per-partition'
```
now need to understand the impact of running it in a long test like this, and for so many days, and see if it won't disturb the whole load.

yarongilor replied:
will start running the heavier longevities soon.

yarongilor replied:
started 4d longevity in: https://jenkins.scylladb.com/job/scylla-staging/job/Longevity_yaron/job/longevity-large-partition-asymmetric-cluster-3h_reversed_queries/40/

yarongilor replied:
Updated results are in:
https://docs.google.com/spreadsheets/d/1L3n-HwSi97_h0tOmDlkB1r69wz9NQYSYDDUXjqnf8YQ/edit?usp=sharing

fgelcer commented on test-cases/longevity/longevity-large-partition-8h.yaml:
```diff
@@ -46,3 +46,5 @@ nemesis_interval: 15
 
 space_node_threshold: 644245094
+
+run_full_partition_scan: '{"ks_cf": "scylla_bench.test", "interval": 2, "pk_name":"pk", "rows_count": 5555}' # 'ks.cf, interval(min), partition-key name, number-of-rows-per-partition'
```
same here

yarongilor commented:
Test now runs ok. only missing part is the result comparison. it is already implemented and commented-out (due to local pc issues  with python 10 + Fedora )
An example event output is:
```
19:40:31  < t:2021-12-05 17:40:31,378 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-12-05 17:40:31.376: (FullPartitionScanReversedOrderEvent Severity.NORMAL) period_type=begin event_id=ed760bfb-fe5c-495c-801c-65b0778601a6 node=longevity-large-partitions-3h-rever-db-node-0a755d00-1 select_from=scylla_bench.test
19:40:31  < t:2021-12-05 17:40:31,437 f:scan_operation_thread.py l:257  c:FullPartitionScanThread p:INFO  > Will run command "select * from scylla_bench.test where pk = 111 and ck > 3068 order by ck desc limit 4857 bypass cache"
19:40:31  < t:2021-12-05 17:40:31,438 f:scan_operation_thread.py l:248  c:FullPartitionScanThread p:INFO  > Will fetch up to 1000 result pages.."
19:40:31  < t:2021-12-05 17:40:31,453 f:scan_operation_thread.py l:299  c:PagedResultHandler   p:INFO  > Will fetch the next page: 0
19:40:31  < t:2021-12-05 17:40:31,469 f:scan_operation_thread.py l:299  c:PagedResultHandler   p:INFO  > Will fetch the next page: 1
19:40:31  < t:2021-12-05 17:40:31,485 f:scan_operation_thread.py l:299  c:PagedResultHandler   p:INFO  > Will fetch the next page: 2
19:40:31  < t:2021-12-05 17:40:31,511 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-12-05 17:40:31.510: (FullPartitionScanReversedOrderEvent Severity.NORMAL) period_type=end event_id=ed760bfb-fe5c-495c-801c-65b0778601a6 duration=0s node=longevity-large-partitions-3h-rever-db-node-0a755d00-1 select_from=scylla_bench.test message=FullPartitionScanReversedOrderEvent operation ended successfully
```
Runs in: https://jenkins.scylladb.com/job/scylla-staging/job/Longevity_yaron/job/longevity-large-partition-asymmetric-cluster-3h_reversed_queries/39/execution/node/142/log/

fgelcer commented on test-cases/longevity/longevity-large-partition-3h.yaml:
```diff
@@ -40,3 +40,7 @@ nemesis_interval: 15
+
+run_fullscan: '{"ks_cf": "scylla_bench.test", "interval": 10}' # 'ks.cf|random, interval(min)'
+
+run_full_partition_scan: '{"ks_cf": "scylla_bench.test", "interval": 2, "pk_name":"pk", "rows_count": 5000, "validate_data": "true"}' # 'ks.cf, interval(min), partition-key name, number-of-rows-per-partition, validate reversed query output'
```
default now for `validate_data` should be `"false"`

yarongilor replied:
fixed

fgelcer reviewed (changes requested):
in addition to the changes below, the commits should be squashed

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,309 @@
+import time
+from abc import abstractmethod
+from typing import Optional
+# from deepdiff import DeepDiff  # TODO: Uncomment when a new hydra docker with deepdiff is available
```
but i want to uncomment it in a follow-up PR soon

fgelcer replied:
as i told you, create a local commit, or even a branch with this code in, but remove from this PR

yarongilor replied:
fixed

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,309 @@
+from cassandra.query import SimpleStatement  # pylint: disable=no-name-in-module
+
+from sdcm.cluster import BaseNode, BaseScyllaCluster, BaseCluster
+# from sdcm.sct_events.system import InfoEvent  # TODO: Uncomment when a new hydra docker with deepdiff is available
```
but i want to uncomment it in a follow-up PR soon

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,309 @@
+        if self.full_partition_scan_params.get('validate_data'):
+            # TODO: remove and uncomment when a new hydra docker with deepdiff is available
+            self.log.debug('Temporarily not executing the normal query of: %s', normal_query)
+            # reversed_query_result = self.query_result_data
+            # ScanOperationThread.run_scan_operation(self, scan_operation_event=scan_operation_event, cmd=normal_query)
+            # diff = DeepDiff(t1=reversed_query_result, t2=self.query_result_data, ignore_order=True,
+            #                 ignore_numeric_type_changes=True)
+            # if diff:
+            #     InfoEvent(
+            #         message=f'Found the following differences between a reversed and normal queries: {diff}').publish()
```
but i want to uncomment it in a follow-up PR soon

yarongilor commented on longevity_test.py:
```diff
@@ -242,11 +238,13 @@ def test_custom_time(self):
-                                                                             fullscan['interval']))
-            self.run_fullscan_thread(ks_cf=fullscan['ks.cf'], interval=fullscan['interval'])
+        fullscan_params = self._get_scan_operation_params(scan_operation='run_fullscan')
+        if fullscan_params:
```
fixed

fgelcer reviewed (commented):
2 small comments, other than that, LGTM

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,300 @@
+    def randomly_add_timeout(cmd) -> str:
+        if random.choice([True] * 2 + [False]):
+            cql_timeout_seconds = str(random.choice([2, 4, 8, 30, 120, 300]))
+            cql_timeout_param = f" USING TIMEOUT {cql_timeout_seconds}s"
```
it is already printed like:
```
< t:2021-12-08 07:46:15,482 f:scan_operation_thread.py l:257  c:FullPartitionScanThread p:INFO  > Will run command "select * from scylla_bench.test where pk = 1535 order by ck desc bypass cache USING TIMEOUT 8s"
< t:2021-12-08 07:47:01,186 f:scan_operation_thread.py l:257  c:FullPartitionScanThread p:INFO  > Will run command "select * from scylla_bench.test where pk = 1535 and ck > 70521 order by ck desc USING TIMEOUT 30s"
```

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,300 @@
+    @staticmethod
+    def randomly_add_timeout(cmd) -> str:
+        if random.choice([True] * 2 + [False]):
+            cql_timeout_seconds = str(random.choice([2, 4, 8, 30, 120, 300]))
```
@fgelcer , this is not a new code of this PR. it is just the old code of the renamed full_scan_thread.py:

https://github.com/scylladb/scylla-cluster-tests/blob/3dd6ef8c6c1386617ec759ebea04f89d7404d676/sdcm/full_scan_thread.py#L48

fgelcer reviewed (approved):
@roydahan , LGTM

yarongilor commented:
@fgelcer , @roydahan , i merged all private branches relevant commits to this main PR.
so it will not pass precommit checks before making a squash. please advice for the content of this list of commits first - is there any to exclude or move to a different PR.

roydahan commented on test-cases/longevity/longevity-large-partition-4days.yaml:
```diff
@@ -2,8 +2,14 @@ test_duration: 6480
 
-prepare_write_cmd:  ["scylla-bench -workload=sequential -mode=write -max-rate=300 -replication-factor=3 -partition-count=10 -clustering-row-count=10000000 -clustering-row-size=5120 -concurrency=7 -rows-per-request=10"]
-stress_cmd: ["scylla-bench -workload=uniform -mode=read -replication-factor=3 -partition-count=10 -clustering-row-count=10000000 -clustering-row-size=5120 -rows-per-request=10 -concurrency=7 -max-rate=32000 -duration=5760m"]
+prepare_write_cmd:  ["scylla-bench -workload=sequential -mode=write -replication-factor=3 -partition-count=20 -clustering-row-count=10000000 -clustering-row-size=2048 -concurrency=50000 -rows-per-request=2000 -timeout=180s -connection-count 10000",
```
Move this to a different PR.

fgelcer replied:
this is in the same PR, but in a different commit (8d1de3b53735e7358f1a130e101c0c46e74d12c3)

yarongilor replied:
opened https://github.com/scylladb/scylla-cluster-tests/pull/4244

roydahan commented on test-cases/longevity/longevity-large-partition-4days.yaml:
```diff
@@ -13,6 +19,8 @@ instance_type_db: 'i3en.3xlarge'
 
 nemesis_class_name: 'SisyphusMonkey'
 nemesis_seed: '014'
+instance_type_loader: 'c5n.9xlarge'
```
This one should be tested with a cheaper loader, we saw that the loader isn't that stressed.

roydahan replied:
c5n.4xlarge should be good enough or even more than enough.

yarongilor replied:
Addressed in opened https://github.com/scylladb/scylla-cluster-tests/pull/4244

fgelcer commented on longevity_test.py:
```diff
@@ -242,11 +245,7 @@ def test_custom_time(self):
-            self.log.info('Fullscan target: {} Fullscan interval: {}'.format(fullscan['ks.cf'],
-                                                                             fullscan['interval']))
-            self.run_fullscan_thread(ks_cf=fullscan['ks.cf'], interval=fullscan['interval'])
+
```
nitpick: why 2 empty lines?

yarongilor replied:
outdated code.

fgelcer commented on sdcm/cluster.py:
```diff
@@ -1732,6 +1732,9 @@ def install_package(self, package_name: str) -> None:
             pkg_cmd = 'apt-get'
+            if package_name == "scylla-manager-agent":
+                # A workaround for: https://github.com/scylladb/scylla-pkg/issues/2578
+                self.remoter.sudo('apt --fix-broken install -y')
```
here you could, instead of running it, just make:
```
pkg_cmd = 'apt --fix-broken'
```

and then having it run on next line (otherwise, you will have it running twice)

yarongilor replied:
it is actually two different commands , following the previous apt-get response recommendation.

roydahan replied:
Please take this out to a separate commit.

fgelcer reviewed (dismissed):
LGTM

roydahan commented on requirements.txt:
```diff
@@ -1,76 +1,88 @@
 #
-# This file is autogenerated by pip-compile
```
@yarongilor I don't think all these edits should be in this file.

yarongilor replied:
file is rolled back to original state.

roydahan commented:
@yarongilor this PR is impossible to review like this.
Please rearrange all your commits, squash whatever is needed together, so I'll review only the relevant commits.

At the moment, I need to review old code that is overwritten later with a newer code.

yarongilor commented on requirements.txt:
```diff
@@ -1,76 +1,88 @@
 #
-# This file is autogenerated by pip-compile
+# This file is autogenerated by pip-compile with python 3.9
```
this is an auto generated file that shouldn't have any impact but i rolled it back anyway.

soyacz replied:
as we are on 3.10, it's better to use 3.10 anyway

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,300 @@
+    # pylint: disable=too-many-arguments
+    def __init__(self, db_cluster: [BaseScyllaCluster, BaseCluster], duration: int, interval: int,
+                 termination_event: threading.Event, page_size: int = 100000, **kwargs):
+        self.ks_cf = kwargs.get('ks_cf')
```
fixed

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,340 @@
+                if random.choice([False] + [True]):  # Randomly add a LIMIT
+                    limit = random.randint(a=rows_count/5, b=rows_count)
+                    query_suffix += f' limit {limit}'
+                if random.choice([False] + [True]):
+                    query_suffix += self.bypass_cache
```
fixed

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,340 @@
+                              error)
+        raise Exception('Failed getting table clustering order from all db nodes')
+
+    def randomly_form_cql_statement(self) -> Optional[tuple[str, str]]:  # pylint: disable=too-many-branches
```
added.

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,340 @@
+
+    @staticmethod
+    def randomly_add_timeout(cmd) -> str:
+        if random.choice([True] * 2 + [False]):
```
it's nice, but why replacing existing code with something more complicated?

soyacz replied:
It was just a tip - this is arguable which one is more complicated.
Generally choices with weights are more explicit with intention that showing True is more probable. Having `[True]*2` may be not so obvious.

yarongilor replied:
This is part of the original full-scan thread that didn't change.
The purpose of it is to test both options - with/without timeout.
i think it was implemented by myself in an old PR.

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,340 @@
+
+    def run_scan_operation(self, scan_operation_event, cmd: str = None):  # pylint: disable=too-many-locals
+        db_node = self.db_node
+        if self.ks_cf == 'random':
```
well, this got a bit complicated, combining old and new code.
i don't want the full-partition thread to choose a random ks_cf if the given one doesn't exist in system. better the thread will fail, indicating a user error of specifying a wrong ks_cf in yaml or something.

soyacz replied:
ok, I get that.

yarongilor commented:
> @yarongilor this PR is impossible to review like this. Please rearrange all your commits, squash whatever is needed together, so I'll review only the relevant commits.
> 
> At the moment, I need to review old code that is overwritten later with a newer code.

squashed all internal commits.

fgelcer reviewed (approved):
LGTM

roydahan commented:
@yarongilor while it's in review, please edit the "commit message" so it won't contain lot of junk messages.

yarongilor commented:
> @yarongilor while it's in review, please edit the "commit message" so it won't contain lot of junk messages.

done.

roydahan commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,352 @@
+    """
+    reversed_query_filter_ck_by = {'lt': ' and {} < {}', 'gt': ' and {} > {}', 'lt_and_gt': ' and {} < {} and {} > {}',
+                                   'no_filter': ''}
+    reversed_query_order = ' order by ck desc'
```
Doesn't the "desc" should be parameter?
Why then do you have get_table_clustering_order?

roydahan replied:
I see below that there is a usage of "self.reversed_order", so probably this is an old piece of code.

yarongilor replied:
removed the old unused line.

roydahan commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,352 @@
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.full_partition_scan_params = kwargs
+        self.full_partition_scan_params['validate_data'] = json.loads(
+            self.full_partition_scan_params.get('validate_data', 'false'))
```
Not clear to me

yarongilor replied:
it converts a string of either 'true' or 'false' to a boolean value.

roydahan replied:
Where do you use it?
I know that the validate data isn't ready yet, but at least I would expect to see it set with 'false' in all the relevant places.

roydahan commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,352 @@
+                query_suffix = limit = ''
+
+                if random.choice([False] + [True]):  # Randomly add a LIMIT
+                    limit = random.randint(a=rows_count/5, b=rows_count)
```
Why 5? just something you picked?
maybe should be picked randomly?

yarongilor replied:
ok changed to "a=1" so it is fully random and can get very small numbers.

roydahan commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,352 @@
+                                                                                         ck_random_max_value,
+                                                                                         ck_name,
+                                                                                         ck_random_min_value)
+                    ck_range = ck_random_max_value - ck_random_min_value - 1  # e.g. 15 - 10 - 1 = 4
+                    if limit and limit < ck_range:
+                        # Example: select * from scylla_bench.test where pk = 1 and ck > 10 and ck < 15
+                        # order by ck desc limit 3
+                        # output of ck should be: [12,13,14]
+                        normal_query += \
+                            f' and {ck_name} < {ck_random_max_value} and {ck_name} >= {ck_random_max_value - limit}'
```
I don't understand why this part is needed.
Why do you care here about the limit and how do the normal vs reversed matters?

yarongilor replied:
It should be because the normal query needs the inversed range. It is actually not really tested now since it requiers the completion of other opened task for that ("DeepDif").
Thus it will only be verified and tested to be correct when the other task is completed.
But the theoretical idea is yet as the example above.

roydahan commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,352 @@
+                # reversed query is: select * from scylla_bench.test where pk = 1 and ck > 10
+                # order by ck desc limit 5
+                # normal query should be: select * from scylla_bench.test where pk = 1 and ck > 15 limit 5
+                if ck_filter == 'lt_and_gt':
```
I think this and the rest of the checks for it aren't that readable, consider using (match-case)

yarongilor replied:
fixed

roydahan commented on sdcm/scan_operation_thread.py:
```diff
@@ -0,0 +1,352 @@
+            else:
+                self.log.info('No partition keys found for table: %s! A reversed query cannot be executed!', self.ks_cf)
+                return None
+        return normal_query, self.randomly_add_timeout(reversed_query)
```
Let's remove the timeout from these queries.

yarongilor replied:
removed.

yarongilor commented on sdcm/cluster.py:
```diff
@@ -3476,6 +3476,26 @@ def verify_node_setup(start_time):
+    :return:
+    """
+    pks_list = []
+    cmd = f'select {pk_name} from {ks_cf} limit {num_of_pk}'
```
outdated.

yarongilor commented on sdcm/cluster.py:
```diff
@@ -3476,6 +3476,26 @@ def verify_node_setup(start_time):
+    """
+    pks_list = []
+    cmd = f'select {pk_name} from {ks_cf} limit {num_of_pk}'
+    cql_result = db_node.run_cqlsh(cmd, split=True, verbose=True)
```
outdated.

yarongilor commented on test-cases/longevity/longevity-large-partition-200k_pks-4days.yaml:
```diff
@@ -69,3 +69,5 @@ user_prefix: 'longevity-large-partitions-200k-pks-4d'
 
 stop_test_on_stress_failure: false
+
+run_full_partition_scan: '{"ks_cf": "scylla_bench.test", "interval": 2, "pk_name":"pk", "rows_count": 100000}' # 'ks.cf, interval(sec), partition-key name, number-of-rows-per-partition'
```
changed.

yarongilor replied:
done

yarongilor commented:
@roydahan,  retested and runs ok after addressing comments in:
https://jenkins.scylladb.com/job/enterprise-2021.1/job/PrivateBuilds/job/ReversedQueries/job/longevity-large-partition-asymmetric-cluster-3h/14/

roydahan commented on test-cases/longevity/longevity-large-partition-4days.yaml:
```diff
@@ -25,3 +25,5 @@ space_node_threshold: 644245094
 table_name: "scylla_bench.test"
 primary_key_column: "pk"
+
+run_full_partition_scan: '{"ks_cf": "scylla_bench.test", "interval": 2, "pk_name":"pk", "rows_count": 10000000}' # 'ks.cf, interval(sec), partition-key name, number-of-rows-per-partition'
```
What about this one?

yarongilor replied:
done

yarongilor commented on sdcm/scan_operation_thread.py:
```diff
@@ -235,42 +235,44 @@ def randomly_form_cql_statement(self) -> Optional[tuple[str, str]]:  # pylint: d
+                        # order by ck desc limit 5
+                        # normal query should be: select * from scylla_bench.test where pk = 1 and ck >= 5 limit 5
+                        reversed_query += self.reversed_query_filter_ck_by[ck_filter].format(ck_name,
+                                                                                             ck_random_min_value)
```
done
